### PR TITLE
[omni, model] feat: add fused MoE forward and EP support for Qwen3-Omni-MoE thinker and update expert weight merging in moe_merge script

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

> Enables fused MoE forward and Expert Parallelism (EP) for the Qwen3-Omni-MoE **thinker** text model, following the same pattern as Qwen3-VL-MoE. The talker is left as a TODO.
>
> The key change is replacing the HF `nn.ModuleList`-based per-expert MLP with a new `Qwen3OmniMoeThinkerTextExperts` module that stores weights as stacked 3-D tensors `(num_experts, out, in)`, compatible with both the fused grouped-GEMM kernel and FSDP2 EP sharding.
>
> Extends `scripts/moe_ckpt_merge/moe_merge.py` to support Qwen3-Omni-MoE checkpoint merging, building on top of the fused MoE / EP PR. The script now auto-detects whether the model is a flat MoE (Qwen3-MoE, DeepSeek) or a nested Omni model (Qwen3-Omni-MoE with thinker/talker sub-configs) and merges expert weights accordingly.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: https://github.com/ByteDance-Seed/VeOmni/pulls?q=moe+merge
- [ ] Format the PR title as `[{modules}] {type}: {description}`

### Test

Validated by running fused moe training on mixed image, video, text datasets.

Validated by running the merge script on a Qwen3-Omni-MoE checkpoint — expert weights are correctly stacked into `(num_experts, out, in)` tensors matching the format expected by `modeling_qwen3_omni_moe.py`.

### API and Usage Example

Set `_moe_implementation = "fused"` in the thinker config to enable the fused path:

```python
config._moe_implementation = "fused"
model = build_foundation_model(config_path=..., ...)
```

```bash
python scripts/moe_ckpt_merge/moe_merge.py \
    --raw_hf_path /path/to/raw_qwen3_omni_moe \
    --merge_hf_path /path/to/merged_qwen3_omni_moe
```

The script auto-detects the model type from the config and handles thinker/talker groups automatically.

### Design & Code Changes

**`modeling_qwen3_omni_moe.py`**

- Add `Qwen3OmniMoeThinkerTextExperts` — replaces `nn.ModuleList` with stacked `nn.Parameter` tensors (`gate_proj`, `up_proj`, `down_proj` each of shape `(num_experts, out, in)`). Supports:
  - `fused_forward`: calls `fused_moe_forward` kernel (training, `_moe_implementation="fused"`)
  - `eager_forward`: loop-per-expert path (inference / eager mode)
  - `_load_from_state_dict`: transparently converts HF per-expert format (`experts.{i}.{proj}.weight`) to stacked format on load, so both original HF and pre-merged checkpoints are accepted
- Patch `Qwen3OmniMoeThinkerTextSparseMoeBlock.__init__` to instantiate `Qwen3OmniMoeThinkerTextExperts` instead of `nn.ModuleList`
- Patch `Qwen3OmniMoeThinkerTextSparseMoeBlock.forward` to pass `routing_weights` and `selected_experts` directly to the experts module
- Patch `Qwen3OmniMoeThinkerForConditionalGeneration.__init__` to propagate `config._moe_implementation` → `config.text_config._moe_implementation` before the text model is built; guarded against multiple calls to `apply_veomni_qwen3_omni_moe_patch()` to prevent infinite recursion

**`parallel_plan.py`**

- Update EP plan key paths from `experts.*.{proj}.weight` (per-expert ModuleList paths) to `experts.{proj}` (stacked Parameter paths), with `Shard(0)` on the expert dimension

**`scripts/moe_ckpt_merge/moe_merge.py`**

- Add `MoEGroup` dataclass — captures `(prefix, num_hidden_layers, num_experts, moe_layer_start_idx)` for one logical group of MoE layers.
- Add `_get_moe_config(config)` — extracts `num_experts` (supports both `num_experts` and `n_routed_experts`) and `first_k_dense_replace` from any config object.
- Add `_detect_moe_groups(config)` — returns a list of `MoEGroup`:
  - Nested Omni models (detected via `thinker_config` / `talker_config`): emits one group per sub-model, skipping talker if it has no MoE layers.
  - Flat models (Qwen3-MoE, DeepSeek): emits a single `model.layers` group, same as the original logic.
- Add `_merge_experts_for_group(state_dict, group)` — consolidates the per-expert loop (`gate_proj`, `up_proj`, `down_proj`) that was previously inlined in `main()`.
- Update `main()` to call `_detect_moe_groups` + `_merge_experts_for_group`; the original flat-model inline loop is superseded but retained for backward compatibility.

### Checklist Before Submitting

- [ ] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md?plain=1#L22): `make commit && make style && make quality`
- [ ] Add / Update [the documentation](https://github.com/ByteDance-Seed/VeOmni/blob/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/ByteDance-Seed/VeOmni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: merge script is validated manually against a real checkpoint; CI integration not feasible without large model artifacts.

### CODE REVIEW

File: veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py [hunk 1]
```diff
@@ -4515,6 +4522,182 @@ def forward(
...
+
+    # --- Patch.1 ---
+
+    def eager_forward(
+        self,
+        hidden_states: torch.Tensor,
+        routing_weights: torch.Tensor,
+        selected_experts: torch.Tensor,
+    ) -> torch.Tensor:
+        """Eager (loop-per-expert) forward – used for inference / non-fused path."""
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py]: do not support EP for eager path? throw an error if ep is enabled and we get in the eager_forward.

File: veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py [hunk 2]
```diff
@@ -4515,6 +4522,182 @@ def forward(
...
+        self,
+        hidden_states: torch.Tensor,
+        routing_weights: torch.Tensor,
+        selected_experts: torch.Tensor,
+    ) -> torch.Tensor:
+        if self.training and self.moe_implementation == "fused":
+            return self.fused_forward(hidden_states, routing_weights, selected_experts)
+        return self.eager_forward(hidden_states, routing_weights, selected_experts)
+
+    def _load_from_state_dict(
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py]: can we remove this?

File: veomni/models/transformers/qwen3_omni_moe/parallel_plan.py [hunk 3]
```diff
@@ -4,13 +4,13 @@
 
 
 def get_parallel_plan():
-    # NOTE: Expert Parallelism (EP) with FSDP2 is NOT ready for this model yet.
-    # This parallel plan is only added to prevent errors during model initialization.
-    # TODO: Implement proper EP support for Qwen3OmniMoe.
+    # Thinker experts use stacked 3-D weight tensors (num_experts, out, in),
+    # so EP shards along dim-0 (the expert dimension).
+    # TODO: Add Talker EP plan once talker fused MoE is implemented.
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3_omni_moe/parallel_plan.py]: nit: can we clarify this that we don't support talker training yet?

File: scripts/moe_ckpt_merge/moe_merge.py [hunk 4]
```diff
@@ -28,6 +28,100 @@ def __iter__(self) -> Generator[Tuple[str, "torch.Tensor"], None, None]:
...
+
+    if hasattr(config, "first_k_dense_replace"):
+        moe_layer_start_idx = config.first_k_dense_replace
+    else:
+        moe_layer_start_idx = 0
+
+    return num_experts, moe_layer_start_idx
+
+
+def _detect_moe_groups(config) -> List[MoEGroup]:
```
Comment (piyifan123) [inline scripts/moe_ckpt_merge/moe_merge.py:56]: check if it's possible to get the groups based on a model specific mapping?  anwway, not a critical comment since we don't need this in transformers v5.

Comment (TimYangst) [reply scripts/moe_ckpt_merge/moe_merge.py:56]: Acknowledge,  let's keep this simple patch at first, I will check if there is a better approach if it is not complex.

File: veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py (hunk 1 above)
Comment (TimYangst) [reply veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py]: Done

File: veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py (hunk 2 above)
Comment (TimYangst) [reply veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py]: Done

File: veomni/models/transformers/qwen3_omni_moe/parallel_plan.py (hunk 3 above)
Comment (TimYangst) [reply veomni/models/transformers/qwen3_omni_moe/parallel_plan.py]: Done

File: veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py [hunk 5]
```diff
@@ -4515,6 +4522,143 @@ def forward(
         return thinker_outputs
 
 
+# ================================================================
+# Patch: Qwen3OmniMoeThinkerTextSparseMoeBlock
+# 1. Replace nn.ModuleList experts with stacked-weight Experts module
+#    to support fused MoE forward and Expert Parallelism (EP)
+# ================================================================
+class Qwen3OmniMoeThinkerTextExperts(nn.Module):
```
Comment (piyifan123) [inline veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py]: can we align this with https://github.com/ByteDance-Seed/VeOmni/blob/main/veomni/models/transformers/qwen3_moe/modeling_qwen3_moe.py#L67

Comment (TimYangst) [reply veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py]: Done
